### DISCUSSION TITLE
Is it possible to search for permits?

### DISCUSSION THREAD

**Discussion by redvving19** [juftin/camply#22] [Q&A] [answered]
Here is an example:
https://www.recreation.gov/permits/233273/registration/detailed-availability?date=2021-06-13

Thanis,

**Comment by juftin** [comment, marked as answer]
@redvving19 great question. Not currently, Recreation.gov doesn't bundle their permit availability and campsite availability together in the same API endpoint. Backcountry permits are theoretically possible, just a bit more challenging to integrate with camply than campsites. This won't be a feature added in the near feature, sorry for the bad news 😢

**Reply by dyurk1113** [reply]
Do you have any pointers as to where I might start if I wanted to try to build a permit-searching extension? I can't find any documentation around the API you access for campground availability (I found RIDB, but that seems to just have general info about trailheads and not availability).

**Reply by juftin** [reply]
Hey @dyurk1113 you're exactly right. The RIDB (https://ridb.recreation.gov) endpoints are all documented. This is where most of the metadata comes from about the various Campsites / Campgrounds / Recreation Areas. The actual campground availability data (and I'm assuming the Permit endpoint) is instead available at the https://recreation.gov/api endpoint. 

I used Chrome Developer tools and [Searched for XHR Network Activity](https://developer.chrome.com/docs/devtools/network/) when interacting with Recreation.gov website to figure out how the https://recreation.gov/api/camps/availability/campground endpoint worked. If there is metadata available via RIDB then the ID field is certainly shared between both. 

Running `camply` via Jupyter Notebooks is my typical exploration method - if it's similar to the Campsite API then the below script might be able to get you started once you find the endpoint, params, etc:

```python
import datetime
import json
from random import choice

from requests import Session

from camply.config import RecreationBookingConfig, STANDARD_HEADERS, USER_AGENTS

session = Session()

headers = STANDARD_HEADERS.copy()
headers.update(choice(USER_AGENTS))
headers.update(RecreationBookingConfig.API_REFERRERS)
session.headers = headers

permit_id = 101
permits_endpoint = f"https://recreation.gov/api/permits/availability/{permit_id}"

search_month = datetime.datetime(2022, 7, 1)
formatted_month = search_month.strftime("%Y-%m-01T00:00:00.000Z")
params = {"start_date": formatted_month}

response = session.get(url=permits_endpoint,
                       params=params,
                       timeout=30)
response_data = json.loads(response.content)
```

Hopefully this is helpful, if you find the endpoint please feel free to follow up here and I can help provide some additional pointers on how to integrate with `camply`

**Reply by juftin** [reply]
Okay, I dug a little deeper.

Here is the RIDB Metadata: https://ridb.recreation.gov/docs#/Permit%20Entrances. "Permit Entrance" is the name the Documentation gives them.

Here's an Example "Permit Entrance", `Arches National Park Canyoneering Permits`: https://www.recreation.gov/permits/4251908 (ID # 4251908). 

Here are some API Endpoints that are hit when browsing for availability:
- https://www.recreation.gov/api/permitdayuse/4251908/availability?year=2022&start_date=2022-01-01&end_date=2022-12-31
    - This looks like the one we want!
- https://www.recreation.gov/api/permitcontent/4251908 
    - This looks like more metadata
- https://www.recreation.gov/api/permitlottery/4251908
- https://www.recreation.gov/api/media/public/asset/4251908
- https://www.recreation.gov/api/search?fq=entity_type%3Arecarea&fq=entity_id%3A2573

Now we just have to find some kind of pattern / structure in the JSON responses and make sense out of them

**Reply by juftin** [reply]
Updated Python Script After Finding the API Endpoint:

```python
import datetime
import json
from pprint import pprint
from random import choice

from requests import Session

from camply.config import RecreationBookingConfig, STANDARD_HEADERS, USER_AGENTS

session = Session()

headers = STANDARD_HEADERS.copy()
headers.update(choice(USER_AGENTS))
headers.update(RecreationBookingConfig.API_REFERRERS)
session.headers = headers

permit_id = 4251908
permits_endpoint = f"https://www.recreation.gov/api/permitdayuse/{permit_id}/availability"

search_date = datetime.date(2022, 7, 1)
start_date = search_date.replace(month=1, day=1)
end_date = search_date.replace(month=12, day=31)

params = dict(start_date=start_date.strftime("%Y-%m-%d"),
              end_date=end_date.strftime("%Y-%m-%d"),
              year=search_date.year)

response = session.get(url=permits_endpoint,
                       params=params,
                       timeout=30)
response_data = json.loads(response.content)

pprint(response_data)
```

**Reply by dyurk1113** [reply]
Thanks for the super detailed reply! Right now I'm particularly looking at Inyo permits, and interestingly the endpoint and params are a bit different:
```
permit_id = 233262
permits_endpoint = f"https://www.recreation.gov/api/permitinyo/{permit_id}/availability

params = dict(start_date=start_date.strftime("%Y-%m-%d"),
              end_date=end_date.strftime("%Y-%m-%d"),
              commercial_acct='false')
```

That returns a list of available/total for all permits in Inyo for the selected date range. Annoyingly they're all labeled with an ID number that doesn't match the listed ID on the website (e.g. Kearsarge Pass has the ID label "JM31" on the website but is just labeled as "465" in the returned availability data). It seems that the "465" number is stored in the RIDB listing for the trailhead, so it shouldn't be too hard to do that mapping.

**Reply by redvving19** [reply]
Blast from the past, If I can help (IE do simple grunt work) let me know.

**Reply by dyurk1113** [reply]
Interestingly, all of the California parks and forests I've checked so far work through the `api/permitinyo` portal. A few sample permit IDs I pulled:

```
# permit_id = 233262 # Inyo NF
# permit_id = 445856 # Humboldt NF
# permit_id = 445857 # Sequoia/Kings Canyon NP
# permit_id = 445858 # Sierra NF
permit_id = 445859 # Yosemite NP
permits_endpoint = f"https://www.recreation.gov/api/permitinyo/{permit_id}/availability"
```

It's weird that there are different API targets for different parks; juftin found `api/permitdayuse` for Canyoneering in Arches, and when I look up Grand Tetons it's just under `api/permits`; I'm not sure if there's any way to find that info consistently other than just checking park by park.

**Reply by dyurk1113** [reply]
Here's my quick-and-dirty version; it's not built out robustly in the same way as the campsite code but it has the basic functionality I need.

Main Script:
```python
import datetime, calendar
import json
from random import choice
from time import sleep, time
from smtplib import SMTP_SSL
from email.message import EmailMessage
import tenacity

from requests import Session

from camply.config import RecreationBookingConfig, STANDARD_HEADERS, USER_AGENTS

WAIT_WINDOW_MIN = 10
MIN_NOTIF_WIN_SEC = 24 * 3600
SEARCH_FILE = 'permit_searches.json'
EMAIL_CONFIG_FILE = 'email_config.json'


# To find more, go to: https://ridb.recreation.gov/docs#/Permit%20Entrances/getFacilityPermitEntrances

def get_reg_url(permit_id, date):
    return f"https://www.recreation.gov/permits/{permit_id}/registration/detailed-availability?date={date}&type=overnight-permit"


@tenacity.retry(wait=tenacity.wait_random_exponential(multiplier=3, max=1800),
                stop=tenacity.stop.stop_after_delay(6000))
def make_request(session, permits_endpoint, params):
    response = session.get(url=permits_endpoint,
                           params=params,
                           timeout=30)
    return response


def run_search_once():
    # DATE RANGES MUST START AND END IN THE SAME MONTH
    with open(SEARCH_FILE, 'r') as f:
        all_permit_data = json.load(f)
    facility_name_id_map = all_permit_data['facility_name_id_map']
    trail_name_id_map = all_permit_data['trail_name_id_map']
    permit_searches = all_permit_data['search_list']

    session = Session()
    headers = STANDARD_HEADERS.copy()
    headers.update(choice(USER_AGENTS))
    headers.update(RecreationBookingConfig.API_REFERRERS)
    session.headers = headers

    available_permits = []
    for ps in permit_searches:
        sd, ed = datetime.date(*ps['start_date']), datetime.date(*ps['end_date'])
        start_date_str = sd.strftime('%Y-%m-%d')
        end_date_str = ed.strftime('%Y-%m-%d')
        start_date_search = sd.replace(day=1).strftime('%Y-%m-%d')
        end_date_search = ed.replace(day=calendar.monthrange(ed.year, ed.month)[1]).strftime('%Y-%m-%d')
        min_party_size = ps['min_party_size']
        params = dict(start_date=start_date_search,
                      end_date=end_date_search,
                      commercial_acct='false')
        for permit_name, trail_name_list in ps['searches'].items():
            permit_id = facility_name_id_map[permit_name]
            permits_endpoint = f"https://www.recreation.gov/api/permitinyo/{permit_id}/availability"
            response = make_request(session, permits_endpoint, params)
            try:
                response_data = json.loads(response.content)['payload']
                # print(response.content)
            except:
                print(response.content)
            for date, date_avails in response_data.items():
                if date < start_date_str or date > end_date_str:
                    continue
                for trail_name in trail_name_list:
                    trail_id = str(trail_name_id_map[trail_name])
                    avail = date_avails[trail_id]
                    rem, tot = avail['remaining'], avail['total']
                    if rem >= min_party_size:
                        reg_url = get_reg_url(permit_id, date)
                        available_permits.append(
                            f"{date}: {rem}/{tot} AVAILABLE FOR {trail_name} ({permit_name})\n\t{reg_url}")
    for ap in available_permits:
        print(ap)
    return available_permits


def run_search_cont(email_server):
    last_pulse_time = 0
    notif_age = {}
    while True:
        available_permits = run_search_once()
        cur_time = time()
        notif_msgs = []
        for ap in available_permits:
            if ap in notif_age and cur_time - notif_age[ap] < MIN_NOTIF_WIN_SEC:
                continue
            notif_age[ap] = cur_time
            notif_msgs.append(ap)
        if len(notif_msgs) > 0:
            send_notifs(email_server, notif_msgs)
            last_pulse_time = cur_time
        sleep(WAIT_WINDOW_MIN * 30)
        if cur_time - last_pulse_time >= MIN_NOTIF_WIN_SEC:
            pulse_resp = send_pulse(email_server)
            if pulse_resp:
                last_pulse_time = cur_time
        sleep(WAIT_WINDOW_MIN * 30)


def send_pulse(email_server):
    with open(EMAIL_CONFIG_FILE, 'r') as f:
        email_config = json.load(f)
    if not email_config['send_daily_pulse']:
        return False
    send_email(email_server,
               "No new permits found today, but I'm still looking!",
               email_config,
               subj_override="Permit Searcher Daily Pulse")
    print('Pulse Sent!')
    return True


def send_notifs(email_server, notif_msgs):
    with open(EMAIL_CONFIG_FILE, 'r') as f:
        email_config = json.load(f)
    message_body = '\n\n'.join(notif_msgs)
    response = send_email(email_server, message_body, email_config)
    if response:
        print('SOMETHING WEIRD HAPPENED WITH EMAIL!')
        print(response)


def send_email(email_server, msg_body, email_config, subj_override=None):
    email = EmailMessage()
    email.set_content(msg_body)
    if subj_override is None:
        email["Subject"] = email_config['subj_line']
    else:
        email["Subject"] = subj_override
    email["From"] = email_config['from_addr']
    email["To"] = ','.join(email_config['to_addrs'])
    response = email_server.send_message(email)
    return response


def email_startup(send_msg=False):
    with open(EMAIL_CONFIG_FILE, 'r') as f:
        email_config = json.load(f)
    email_server = SMTP_SSL(email_config['smtp_server'],
                            email_config['smtp_port'])
    email_server.ehlo()
    email_server.login(user=email_config['email_username'],
                       password=email_config['email_password'])

    if send_msg:
        resp = send_email(email_server, 'Testing!!!!', email_config)
        print(resp)

    return email_server


if __name__ == '__main__':
    # Test email config
    email_server = email_startup(False)
    run_search_cont(email_server)

```

Email Config:
```json
{
  "to_addrs": ["<enter_emails>"],
  "from_addr": "<enter_email>",
  "subj_line": "PERMIT AVAILABILITY DETECTED",
  "smtp_server": "smtp.gmail.com",
  "email_username": "<enter_username>",
  "email_password": "<enter_pw>",
  "smtp_port": 465,
  "send_daily_pulse": true
}
```

Search Config (not nearly exhaustive, but should have most of the most competitive permits in each park):
```json
{
  "search_list": [
    {
      "start_date": [2022, 8, 5],
      "end_date": [2022, 8, 6],
      "min_party_size": 3,
      "searches": {
        "Inyo": [
          "Kearsarge Pass"
        ]
      }
    }
  ],
  "facility_name_id_map": {
    "Inyo": 233262,
    "SeKi": 445857,
    "Yosemite": 445859,
    "Sierra": 445858,
    "Humboldt": 445856,
    "Half Dome Cables": 234652,
    "Whitney": 233260
  },
  "trail_name_id_map": {
    "----------Inyo SECTION----------": -1,
    "Baxter Pass": 464,
    "Big Pine Creek North Fork": 495,
    "Big Pine Creek South Fork": 461,
    "Bishop Pass": 459,
    "Cottonwood Lakes": 520,
    "Cottonwood Pass": 440,
    "Deer Lakes": 443,
    "Duck Pass": 444,
    "High Trail": 485,
    "Hilton Lakes": 449,
    "Italy Pass": 452,
    "JMT North": 486,
    "JMT South": 476,
    "Kearsarge Pass": 465,
    "Lamarck Lakes": 446,
    "Little Lakes Valley": 451,
    "Minaret Lake": 436,
    "Mono Pass": 494,
    "North Fork of Lone Pine Creek": 512,
    "Piute Pass": 456,
    "Sabrina Lake": 482,
    "Shepherd Pass": 466,
    "Treasure Lakes": 458,
    "----------SEKI SECTION----------": -1,
    "Bubbs Creek": 44585704,
    "Copper Creek": 44585702,
    "Franklin": 44585724,
    "High Sierra Trail": 44585711,
    "Sawtooth Pass": 44585717,
    "Timber Gap": 44585716,
    "Twin Lakes": 44585732,
    "----------YOSEMITE SECTION----------": -1,
    "Beehive Meadows": 44585904,
    "Cathedral Lakes": 44585907,
    "Glen Aulin": 44585914,
    "Glen Aulin Passthrough": 44585915,
    "Happy Isles -> LYV": 44585917,
    "Happy Isles -> Past LYV": 44585918,
    "Lyell Canyon": 44585921,
    "May Lake": 44585924,
    "Rafferty Creek -> Vogelsang": 44585940,
    "Sunrise Lakes": 44585945,
    "Young Lakes via Glen Aulin": 44585956,
    "----------HALF DOME SECTION----------": -1,
    "Half Dome Cables": 31,
    "----------WHITNEY SECTION----------": -1,
    "Whitney": 215
  }
}
```

**Reply by juftin** [reply]
Whoa, this is impressive @dyurk1113. Really nice work! I'm wondering about a few things here: 

- Is the `permitinyo` API endpoint exclusive to permits underneath the [Inyo National Forest Recreation Area](https://www.recreation.gov/camping/gateways/1064)
    - In particular I'm wondering about just how many variations of these permit endpoints there are. Is there one endpoint to rule them all that we might be able to use across all permits, or is there somewhere that gives us the metadata of the endpoint to hit? I found the below endpoint while playing around a bit with Half Dome Cables.
        - `https://recreation.gov/api/permits/{facility_id}/availability/month?start_date=2022-03-01T00:00:00.000Z`

Full transparency, I haven't gotten to sit down and really dig into this yet. I also have a bit of a knowledge gap of how permits work through recreation.gov, what different kinds of permits there are, and which scenarios is camply a good fit for. I will have some more to follow up here as get an opportunity to dig into this further. But I really appreciate all of this, I'm excited about building some new functionality.

**Reply by dyurk1113** [reply]
I've tested that `permitinyo` at least works for Inyo NF, Sierra NF, Humboldt NF, Yosemite NP, and Sequoia/Kings Canyon NP (each of which has its own booking page on the rec.gov website), but it would be nice to find some sort of master entry point since `permitinyo` likely won't work all over the country. Maybe just using `permits` will work for other place as well as the Half Dome Cables.

In terms of use case, the main way I wanted to adapt your code for permits is almost the exact same as what you've done for campsites. Most popular permits are released ~6 months in advance (at least in the California areas I go to) and the quotas get booked up almost immediately, but if someone cancels after that the spots open up briefly if you're quick enough to snag them. In some ways it should be a bit simpler than campsites, because a) you just have to check one quota for each trail rather than checking different campsites within a larger campground and b) wilderness permits are only based on your entry date and don't carry any time limit for how long you can be out, so you only have to check if one entry date is open rather than looking for continuous date ranges.

The main challenge I've hit so far is figuring out the hidden ID numbers for each trailhead without manually scanning network packets. For some reason the IDs are easy to find in RIDB for Inyo permits, but I can't find them on RIDB for Yosemite or Sequoia/Kings Canyon.

**Reply by alexblanck** [reply]
FWIW, the availability for a few permits I'm interested in seems to be at a similar endpoint to the one @juftin mentioned above. Here's an example:

```
https://www.recreation.gov/api/permits/4675309/availability?start_date=2022-05-16T07%3A00%3A00.000Z&end_date=2022-06-30T00%3A00%3A00.000Z&commercial_acct=false&is_lottery=false
```

As far as finding the facility/permit ids, I think @dyurk1113 is right that this might be the more difficult part.

**Reply by dchaniel** [reply]
Hi! Thanks @redvving19 for the detailed code, planning to try that out, and I'd love to add it as a feature for camply if there's a robust way to do it. Do you or @juftin have any updates from the last year on permit searching? If not I'll start from the great example you left in this thread.

**Reply by juftin** [reply]
Hey @dchaniel. @changlan has done most of the hard work here: https://github.com/juftin/camply/pull/281

I owe them some code review. The new feature would need some documentation and testing as well

**Comment by redvving19** [comment]
No problem, thought two birds with one stone as permits (such as half dome) are very hard to get but people cancel last minute and they become available. 

Great work

**Reply by juftin** [reply]
Thank you. I have seen some kind of support for this over at https://campnab.com. It's not free, but a good service all the same. Good luck!

**Comment by redvving19** [comment]
Here is what I was hoping to solve:

" The walk-up lottery at the Leavenworth Ranger Station is suspended for the 2022 season. Walk-up permits will be released 1 week at a time randomly on Sundays. "

somehow alert when permits are released in order to snag one

**Reply by dyurk1113** [reply]
I believe something like the code I posted would work for that use case, assuming the availability is updated on rec.gov in the same way. However, it would only alert you when it happens to check with whatever frequency it makes requests (default 10 minutes); if the permits already got booked in that window you'd be out of luck. My guess is they probably do their best to block anything that would notify you instantly when permits appear in order to combat bots.

**Comment by redvving19** [comment]
Interesting enough I stumbled across another website charging a fee for this service on reddit:

https://outdoorstatus.com/last-issue/
